payjoin/src/send/multi_party/mod.rs

```diff
@@ -0,0 +1,280 @@
+        let reply_key = self.hpke_ctx.reply_pair.secret_key();
+        let body = serialize_v2_body(&self.psbt, false, None, FeeRate::BROADCAST_MIN)?;
+        let mut ohttp_keys = self
+            .directory_url
+            .ohttp()
+            .map_err(|_| InternalCreateRequestError::MissingOhttpConfig)?;
+        let (request, ohttp_ctx) = extract_request(
+            ohttp_relay,
+            reply_key.clone(),
+            body,
+            self.directory_url.clone(),
+            self.hpke_ctx.receiver.clone(),
+            &mut ohttp_keys,
+        )
+        .map_err(InternalCreateRequestError::V2CreateRequest)?;
+        Ok((request, ohttp_ctx))
+    }
+
+    pub fn process_response(
+        self,
+        response: &[u8],
+        ohttp_ctx: ohttp::ClientResponse,
+    ) -> Result<(), FinalizeResponseError> {
+        let response_array: &[u8; crate::directory::ENCAPSULATED_MESSAGE_BYTES] = response
+            .try_into()
+            .map_err(|_| InternalFinalizeResponseError::InvalidSize(response.len()))?;
+
+        let response = ohttp_decapsulate(ohttp_ctx, response_array)
+            .map_err(InternalFinalizeResponseError::Ohttp)?;
+        match response.status() {
+            http::StatusCode::OK | http::StatusCode::ACCEPTED => Ok(()),
+            _ => Err(InternalFinalizeResponseError::UnexpectedStatusCode(response.status()))?,
+        }
+    }
+}
+
+pub(crate) struct PsbtContext {
+    inner: crate::send::PsbtContext,
+}
+
+impl PsbtContext {
+    fn process_proposal(self, mut proposal: Psbt) -> InternalResult<Psbt> {
+        // TODO(armins) add multiparty check fees modeled after crate::send::PsbtContext::check_fees
+        // The problem with this is that some of the inputs will be missing witness_utxo or non_witness_utxo field in the psbt so the default psbt.fee() will fail
+        // Similarly we need to implement a check for the inputs. It would be useful to have all the checks as crate::send::PsbtContext::check_inputs
+        // However that method expects the reciver to have provided witness for their inputs. In a ns1r the reciever will not sign any inputs of the optimistic merged psbt
+        self.inner.basic_checks(&proposal)?;
+        self.inner.check_outputs(&proposal)?;
+        self.inner.restore_original_utxos(&mut proposal)?;
+        Ok(proposal)
+    }
+}
+
+fn append_optimisitic_merge_query_param(url: &mut Url) {
+    url.query_pairs_mut().append_pair("optimisticmerge", "true");
+}
+
+#[cfg(test)]
+mod test {
+    use std::str::FromStr;
+
+    use bitcoin::{FeeRate, ScriptBuf};
+    use url::Url;
+
+    use crate::send::multi_party::append_optimisitic_merge_query_param;
+    use crate::send::{serialize_url, v1};
+    use crate::HpkeKeyPair;
+
+    #[test]
+    fn req_ctx_ser_de_roundtrip() {
+        use super::*;
+        use crate::send::test::ORIGINAL_PSBT;
+        let v2_sender = v2::Sender {
+            v1: v1::Sender {
+                psbt: Psbt::from_str(ORIGINAL_PSBT).unwrap(),
```

Comment:
We are trying to [Remove `.unwrap()` even in tests](https://github.com/payjoin/rust-payjoin/issues/487) but this can be a follow up

Reply:
Oh yeah. Let me just make that change here

Reply:
Done